### PR TITLE
Fix the comment of  Translation Matrix

### PR DESCRIPTION
**Pull request by robotcator**

### CODE REVIEW

File: gensim/models/translation_matrix.py [hunk 1]
```diff
@@ -139,29 +137,35 @@ def __init__(self, word_pair, source_lang_vec, target_lang_vec, random_state=Non
...
+
+    def train(self, word_pairs):
         """
-        build the translation matrix that mapping from source space to target space.
+        Build the translation matrix that mapping from source space to target space.
 
         Args:
-            `word_pair` (list): a list pair of words
+            `word_pairs` (list): a list pair of words
             `source_space` (Space object): source language space
```
Comment (menshikh-iv) [inline gensim/models/translation_matrix.py]: train method use only `word_pairs`, what is source/target space here?

File: gensim/models/translation_matrix.py [hunk 2]
```diff
@@ -139,29 +137,35 @@ def __init__(self, word_pair, source_lang_vec, target_lang_vec, random_state=Non
...
             `target_space` (Space object): target language space
 
         Returns:
             `translation matrix` that mapping from the source language to target language
         """
-        self.source_space = Space.build(self.source_lang_vec, set(self.source_word))
-        self.target_space = Space.build(self.target_lang_vec, set(self.target_word))
+        self.source_word, self.target_word = zip(*word_pairs)
+        if self.translation_matrix is None:
```
Comment (menshikh-iv) [inline gensim/models/translation_matrix.py]: But if I called train twice, in the second time I don't fit model.
Please remove this `if`

File: gensim/test/test_translation_matrix.py [hunk 3]
```diff
@@ -25,64 +25,57 @@ def temp_save_file():
...
 class TestTranslationMatrix(unittest.TestCase):
     def setUp(self):
-        self.train_file = datapath("OPUS_en_it_europarl_train_one2ten.txt")
-
         self.source_word_vec_file = datapath("EN.1-10.cbow1_wind5_hs0_neg10_size300_smpl1e-05.txt")
         self.target_word_vec_file = datapath("IT.1-10.cbow1_wind5_hs0_neg10_size300_smpl1e-05.txt")
 
-        with utils.smart_open(self.train_file, "r") as f:
-            self.word_pair = [tuple(utils.to_unicode(line).strip().split()) for line in f]
+        self.word_pairs = [("one", "uno"), ("two", "due"), ("three", "tre"),
```
Comment (menshikh-iv) [inline gensim/test/test_translation_matrix.py]: Please use hanging indents

File: gensim/test/test_translation_matrix.py [hunk 4]
```diff
@@ -25,64 +25,57 @@ def temp_save_file():
...
+                           ("four", "quattro"), ("five", "cinque"), ("six", "sei"),
+                           ("seven", "sette"), ("eight", "otto"), ("nine", "nove"),
+                           ("dog", "cane"), ("pig", "maiale"), ("fish", "cavallo"),
+                           ("birds", "uccelli"), ("apple", "mela"), ("orange", "arancione"),
+                           ("grape", "acino"), ("banana", "banana"), ("mango", "mango")
+        ]
+
+        self.test_word_pairs = [("ten", "dieci"), ("dog", "cane"), ("cat", "gatto")]
```
Comment (menshikh-iv) [inline gensim/test/test_translation_matrix.py]: Remove `("dog", "cane")` from `self.word_pairs`

File: gensim/models/translation_matrix.py [hunk 5]
```diff
@@ -91,7 +91,7 @@ def build(cls, lang_vec, lexicon=None):
         return Space(mat, words)
 
     def normalize(self):
-        """ normalized the word vector's matrix """
+        """ Normalized the word vector's matrix """
```
Comment (gojomo) [inline gensim/models/translation_matrix.py]: 'Normalize…' (imperative rather than past-tense)

File: gensim/models/translation_matrix.py [hunk 6]
```diff
@@ -173,13 +175,13 @@ def save(self, *args, **kwargs):
...
     def load(cls, *args, **kwargs):
-        """ load the pre-trained translation matrix model"""
+        """ Load the pre-trained translation matrix model"""
         model = super(TranslationMatrix, cls).load(*args, **kwargs)
         return model
 
     def apply_transmat(self, words_space):
         """
-        mapping the source word vector to the target word vector using translation matrix
+        Mapping the source word vector to the target word vector using translation matrix
```
Comment (gojomo) [inline gensim/models/translation_matrix.py]: 'Map…' (imperative rather than '-ing' form)

**Comment by gojomo** [comment]
The handling of `word_pairs` in `__init__()` and `train()` now makes sense, thanks. The comments have been improved but still may benefit from a deep review for clarity/wording. 

Though I know I requested the Doc2Vec-related example, in its current form the motivations/benefits are muddled. Really it shouldn't require a separate helper class (`BackMappingTranslationMatrix`), and the notebook section ("Tranlation Matrix Revisit") is hard-to-follow, and includes a bunch of improper practices. (For example: using an imbalanced set of docs for the mapping 'overlap'; using the slow-and-iffy `dm_concat` mode; calling `train()` multiple times with a sawtooth `alpha` progression, etc.) 

The word-translation example can presumably be evaluated based on real datasets in the original context that motivated the approach, while the doc-vec example will need more novel design/evaluation – so I'd recommend splitting them into separate notebooks.

**Comment by robotcator** [comment]
Thanks. You do remind me the imbalanced set problem in the example. And the code for training document vector are borrowed from the `doc2vec-imdb.ipynb` and I will re-train the document vector.

For the `BackMappingTranslationMatrix` class, I didn't find a good way to integrate this function into my  `TranslationMatrix` class, so I separate this into two class.

I didn't catch that `The word-translation example can be evaluated based on real datasets in the original context `, can you please explain in more detail?

**Comment by gojomo** [comment]
I meant there are published papers about using word-vector transformations for language-translation - the original Google paper, the Dinu paper – so there are specific datasets & procedures to mimic – and similar results would indicate everything is working. The Doc2Vec use is novel so requires more experimentation/thought.

**Comment by robotcator** [comment]
The word pairs used in this experiment are extracted from the OPUS(http://opus.lingfil.uu.se/). The same as the [Ninu's paper](https://arxiv.org/pdf/1412.6568.pdf). I plot the vis to show the linear relationship between two language vector space and use the word translation to show this transformation works. More re-produced experiments from the [Mikolov's](https://arxiv.org/pdf/1309.4168.pdf) and [Ninu's](https://arxiv.org/pdf/1412.6568.pdf) paper would be fine to support this transformation. But I still can not find any experiment for language translation(Do it mean sentences translation ) from the two paper I mentioned here? Can you remind me If I miss something?

I added "unstable/experimental" warning tag in notebook explicitly for `doc2vec` transformation part.

I also found a paper [(OFFLINE  BILINGUAL  WORD  VECTORS,ORTHOGONALTRANSFORMATIONS AND  THE INVERTED SOFTMAX) ](https://arxiv.org/pdf/1702.03859.pdf) is related to this experiment. But I‘m just having a look and need dig deeper.

**Comment by menshikh-iv** [comment]
Thank you @robotcator for fast fixes.
Thank you @gojomo for a review:+1:

**Comment by menshikh-iv** [comment]
Need to fix some typos/pep8 issues in a notebook, but I can't wait for more, it's release time.

**Pull request merged**